src/DataStructures/VectorImpl.hpp

```diff
@@ -0,0 +1,410 @@
+  explicit VectorImpl(size_t set_size) noexcept
+      : owned_data_(set_size > 0 ? static_cast<value_type*>(
+                                       malloc(set_size * sizeof(value_type)))
+                                 : nullptr,
+                    &free) {
+#if defined(SPECTRE_DEBUG) || defined(SPECTRE_NAN_INIT)
+    std::fill(owned_data_.get(), owned_data_.get() + set_size,
+              std::numeric_limits<value_type>::signaling_NaN());
+#endif  // SPECTRE_DEBUG
+    reset_pointer_vector(set_size);
+  }
+
+  VectorImpl(size_t set_size, double value) noexcept
+      : owned_data_(set_size > 0 ? static_cast<value_type*>(
+                                       malloc(set_size * sizeof(value_type)))
+                                 : nullptr,
+                    &free) {
+    std::fill(owned_data_.get(), owned_data_.get() + set_size, value);
+    reset_pointer_vector(set_size);
+  }
+  // @}
+
+  /// Create a non-owning VectorImpl that points to `start`
+  VectorImpl(T* start, size_t set_size) noexcept
+      : BaseType(start, set_size), owning_(false) {}
+
+  /// Create from an initializer list of doubles. All elements in the
+  /// `std::initializer_list` must have decimal points
+  template <class U, Requires<cpp17::is_same_v<U, T>> = nullptr>
+  VectorImpl(std::initializer_list<U> list) noexcept
+      : owned_data_(list.size() > 0 ? static_cast<value_type*>(malloc(
+                                          list.size() * sizeof(value_type)))
+                                    : nullptr,
+                    &free) {
+    // Note: can't use memcpy with an initializer list.
+    std::copy(list.begin(), list.end(), owned_data_.get());
+    reset_pointer_vector(list.size());
+  }
+
+  /// Empty VectorImpl
+  VectorImpl() = default;
+  /// \cond HIDDEN_SYMBOLS
+  ~VectorImpl() = default;
+
+  // clang-tidy: calling a base constructor other than the copy constructor.
+  // We reset the base class in reset_pointer_vector after calling its default
+  // constructor
+  VectorImpl(const VectorImpl<T, VectorType>& rhs) noexcept;
+
+  VectorImpl(VectorImpl<T, VectorType>&& rhs) noexcept;
+
+  // This is a converting constructor. clang-tidy complains that it's not
+  // explicit, but we want it to allow conversion.
+  // clang-tidy: mark as explicit (we want conversion to VectorImpl type)
+  template <
+      typename VT, bool VF,
+      Requires<cpp17::is_same_v<typename VT::ResultType, VectorType>> = nullptr>
+  VectorImpl(const blaze::DenseVector<VT, VF>& expression) noexcept;  // NOLINT
+
+  template <typename VT, bool VF>
+  VectorImpl& operator=(const blaze::DenseVector<VT, VF>& expression) noexcept;
+
+  //  declaration for copy constructor
+  VectorImpl& operator=(const VectorImpl<T, VectorType>& rhs) noexcept;
+
+  // declaration for move constructor
+  VectorImpl& operator=(VectorImpl<T, VectorType>&& rhs) noexcept;
+  /// \endcond
+
+  // The case of assigning a type apart from the same VectorImpl or a
```

Comment:
Either this comment needs to be promoted to a doxygen comment, or it should be moved to the definition